hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AddMigrationListenerMessageTask.java

```diff
@@ -0,0 +1,145 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.impl.protocol.task;
+
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.client.impl.protocol.codec.ClientAddMigrationListenerCodec;
+import com.hazelcast.client.impl.proxy.PartitionServiceProxy;
+import com.hazelcast.cluster.Member;
+import com.hazelcast.instance.impl.Node;
+import com.hazelcast.internal.nio.Connection;
+import com.hazelcast.internal.partition.IPartitionService;
+import com.hazelcast.partition.MigrationListener;
+import com.hazelcast.partition.MigrationState;
+import com.hazelcast.partition.ReplicaMigrationEvent;
+
+import javax.annotation.Nullable;
+import java.security.Permission;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+import static com.hazelcast.internal.partition.InternalPartitionService.MIGRATION_EVENT_TOPIC;
+import static com.hazelcast.internal.partition.InternalPartitionService.MIGRATION_EVENT_TOPIC_ORDER_KEY;
+import static com.hazelcast.spi.impl.InternalCompletableFuture.newCompletedFuture;
+
+public class AddMigrationListenerMessageTask
+        extends AbstractAddListenerMessageTask<Boolean> {
+
+    public AddMigrationListenerMessageTask(ClientMessage clientMessage, Node node, Connection connection) {
+        super(clientMessage, node, connection);
+    }
+
+    @Override
+    protected CompletableFuture<UUID> processInternal() {
+        IPartitionService partitionService = getService(getServiceName());
+
+        MigrationListener listener = createMigrationListener();
+
+        if (parameters) {
+            return newCompletedFuture(partitionService.addLocalMigrationListener(listener));
+        }
+
+        return partitionService.addMigrationListenerAsync(listener);
+    }
+
+    private MigrationListener createMigrationListener() {
+        return new MigrationListener() {
+
+                @Override
+                public void migrationStarted(MigrationState state) {
+                    sendIfAlive(encodeMigrationEvent(state, PartitionServiceProxy.MIGRATION_STARTED));
+                }
+
+                @Override
+                public void migrationFinished(MigrationState state) {
+                    sendIfAlive(encodeMigrationEvent(state, PartitionServiceProxy.MIGRATION_FINISHED));
+                }
+
+                @Override
+                public void replicaMigrationCompleted(ReplicaMigrationEvent event) {
+                    sendIfAlive(encodeReplicaMigrationEvent(event));
+                }
+
+                @Override
+                public void replicaMigrationFailed(ReplicaMigrationEvent event) {
+                    sendIfAlive(encodeReplicaMigrationEvent(event));
+                }
+            };
+    }
+
+    private void sendIfAlive(ClientMessage eventMessage) {
+        if (endpoint.isAlive()) {
+            sendClientMessage(MIGRATION_EVENT_TOPIC_ORDER_KEY, eventMessage);
```

Comment:
Shouldn't we call `sendClientMessage(null, eventMessage);` here? `MIGRATION_EVENT_TOPIC_ORDER_KEY` constant is not a partition id. @sancar WDYT?

Reply:
@peterjot I'm afraid that I forced you to revert a properly working code with this comment. 😅

I've discussed this call with @sancar and it appears that we need to have a partition id (or a magic number to be more precise) here. That's because of the following.

When the message is send this way, `-1` is [implicitly used](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractMessageTask.java#L263) as the partition id. Then the Java client [receives](https://github.com/hazelcast/hazelcast/blob/eb8ffc40b2e2d7ccfa21e2acfd663bfff1496c6b/hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnection.java#L242) an event, it sees the `-1` and executes the corresponding handler on [a random thread](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/listener/ClientListenerServiceImpl.java#L167). Which means that we're loosing the order on the client side when we use `sendClientMessage(null, eventMessage);` here.

Could you restore the previous implementation and add a comment around this call to help with avoiding confusion in the future?

Reply:
Sure, I also added a test to check it :P

Reply:
Thanks. Let me run the test suite now.

Reply:
Could you also add a `use a constant "partition id" to enforce event handling order` comment here?
https://github.com/hazelcast/hazelcast/pull/17713/files#diff-98463052a544c8d2833cd2c4761dad8b0187e5653e20b84452bad5ecf3a636d6R85